README.md
# timeprofile

This package provides a module for time profiling emissions output.

***This software is provided for research purposes only. It's output may
not accurately reflect observed data due to numerous reasons. Data are
provisional; use at own risk.***

## Python 2 and 3 Support

This package was originally developed to support python 2.7, but has since
been refactored to support 3.5. Attempts to support both 2.7 and 3.5 have
been made but are not guaranteed.

## Development

### Clone Repo

Via ssh:

    git clone git@github.com:pnwairfire/timeprofile.git

or http:

    git clone https://github.com/pnwairfire/timeprofile.git

### Install Dependencies

Run the following to install dependencies:

    pip install -r requirements.txt

Run the following to install packages required for development and testing:

    pip install -r requirements-test.txt
    pip install -r requirements-dev.txt

### Setup Environment

To import timeprofile in development, you'll have to add the repo root directory
to the search path.

## Running tests

Use pytest:

    py.test

You can also use the ```--collect-only``` option to see a list of all tests.

    py.test --collect-only

See [pytest](http://pytest.org/latest/getting-started.html#getstarted) for more information about

## Installing

### Installing With pip

TODO: ...Fill In...

First, install pip (with sudo if necessary):

    apt-get install python-pip
    pip install --upgrade pip

Then, to install, for example, v1.1.2, use the following (with sudo if necessary):

    pip install --extra-index https://pypi.airfire.org/simple timeprofile==1.1.2

If you get an error like

    AttributeError: 'NoneType' object has no attribute 'skip_requirements_regex

it means you need in upgrade pip.  One way to do so is with the following:

    pip install --upgrade pip

## Usage

The simplest case is when the time window is a 24-hour period,
from midnight to midnight.  In this case, time profiling would yield the
following:

    ...FILL IN...

A more complicated scenario would be a time window of, say, 36 hours, starting at 6:00am on day and going until 6:00pm the following day.  Time profiling would yield the following:

    ...FILL IN...

A even more complicated scenario would be a time window of, say, 36.75 hours, starting at 6:30am on day and going until 7:15pm the following day.  Time profiling would yield the following:

    ...FILL IN...


changelog.md

## 1.1.0
 - imeplement FEPS time profile model, with new class `FepsTimeProfiler`
 - add dev script `feps-plotter`

## 1.1.1
 - remove pyairfire dependency

## 1.1.2
 - bug fix and added data validation


test/timeprofile/test_feps.py
__author__      = "Joel Dubowy"

#import copy
import datetime

from numpy.testing import assert_approx_equal
from py.test import raises

from timeprofile.static import FepsTimeProfiler

def assert_approximately_equal(expected, actual):
    if isinstance(expected, dict) and isinstance(actual, dict):
        assert set(expected.keys()) == set(actual.keys())
        for k in expected:
            assert_approximately_equal(expected[k], actual[k])
    elif isinstance(expected, list) and isinstance(actual, list):
        assert len(expected) == len(actual)
        for i in range(len(expected)):
            assert_approximately_equal(expected[i], actual[i])
    elif isinstance(expected, float) and isinstance(actual, float):
        assert_approx_equal(actual, expected, significant=8)  # arbitrarily chose 8
    else:
        assert expected == actual

class TestFepsTimeProfiler_Rx(object):

    def test_three_hours(self):
        s = datetime.datetime(2015, 1, 1, 9, 0)
        e = datetime.datetime(2015, 1, 2, 12, 0)
        stp = FepsTimeProfiler(s, e)

        # TDOD: add assert to check hourly fractions


test/timeprofile/test_static.py
__author__      = "Joel Dubowy"

#import copy
import datetime

from numpy.testing import assert_approx_equal
from py.test import raises

from timeprofile.static import (
    StaticTimeProfiler,
    InvalidStartEndTimesError,
    InvalidHourlyFractionsError
)

def assert_approximately_equal(expected, actual):
    if isinstance(expected, dict) and isinstance(actual, dict):
        assert set(expected.keys()) == set(actual.keys())
        for k in expected:
            assert_approximately_equal(expected[k], actual[k])
    elif isinstance(expected, list) and isinstance(actual, list):
        assert len(expected) == len(actual)
        for i in xrange(len(expected)):
            assert_approximately_equal(expected[i], actual[i])
    elif isinstance(expected, float) and isinstance(actual, float):
        assert_approx_equal(actual, expected, significant=8)  # arbitrarily chose 8
    else:
        assert expected == actual

class TestStaticTimeProfiler_ValidationMethods(object):

    def _dummy_profiler(self, monkeypatch):
        monkeypatch.setattr(StaticTimeProfiler, "_compute_hourly_fractions",
            lambda *args: None)
        return StaticTimeProfiler(None, None)

    def test_validate_start_end_times(self, monkeypatch):
        profiler = self._dummy_profiler(monkeypatch)

        st = datetime.datetime(2015, 1, 1, 11)
        et = datetime.datetime(2015, 1, 1, 12)
        # start can't come after end
        with raises(InvalidStartEndTimesError) as e:
            profiler._validate_start_end_times(et, st)
        # start can't equal end
        with raises(InvalidStartEndTimesError) as e:
            profiler._validate_start_end_times(st, st)
        # proper order of dates shouldn't raise error
        profiler._validate_start_end_times(st, et)

    # TODO: add test for _validate_hourly_fractions

class TestStaticTimeProfiler_DefaultHourlyFractions(object):


    def test_one_day(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 2, 0)
        stp = StaticTimeProfiler(st, et)
        #assert set(stp.FIELDS) == set(stp.hourly_fractions.keys())
        expected_hourly_fractions = {
            p: stp.DEFAULT_DAILY_HOURLY_FRACTIONS[p] for p in stp.FIELDS
        }
        assert_approximately_equal(expected_hourly_fractions, stp.hourly_fractions)
        assert stp.start_hour == st
        assert stp.end_hour == datetime.datetime(2015, 1, 1, 23)

    def test_two_days(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 3, 0)
        stp = StaticTimeProfiler(st, et)
        #assert set(stp.FIELDS) == set(stp.hourly_fractions.keys())
        expected_hourly_fractions = {
            p: 2 * map( lambda e: e / 2.0, stp.DEFAULT_DAILY_HOURLY_FRACTIONS[p])
                for p in stp.FIELDS
        }
        assert_approximately_equal(expected_hourly_fractions, stp.hourly_fractions)
        assert stp.start_hour == st
        assert stp.end_hour == datetime.datetime(2015, 1, 2, 23)

    def test_partial_days(self):
        s = datetime.datetime(2015, 1, 1, 12, 20)
        e = datetime.datetime(2015, 1, 2, 16, 40)
        stp = StaticTimeProfiler(s, e)
        # TDOD: add assert to check hourly fractions
        assert stp.start_hour == datetime.datetime(2015, 1, 1, 12)
        assert stp.end_hour == datetime.datetime(2015, 1, 2, 16)


class TestStaticTimeProfiler_CustomDailyHourlyFractions(object):

    DAILY_HOURLY_FRACTIONS = {
        'area_fraction': [
            0.004, 0.006, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.10, 0.10, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
        ],
        'flaming': [
            0.005, 0.005, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.10, 0.10, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
        ],
        'residual': [  # shifted 6 hrs
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
            0.005, 0.005, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.10, 0.10, 0.05, 0.05, 0.05, 0.05,
        ],
        'smoldering': [  # shifted another 6 hrss
            0.10, 0.10, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
            0.005, 0.005, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
        ]
    }

    def test_one_day(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 2, 0)
        stp = StaticTimeProfiler(st, et, hourly_fractions=self.DAILY_HOURLY_FRACTIONS)
        #assert set(stp.FIELDS) == set(stp.hourly_fractions.keys())
        expected_hourly_fractions = {
            p: self.DAILY_HOURLY_FRACTIONS[p] for p in stp.FIELDS
        }
        assert_approximately_equal(expected_hourly_fractions, stp.hourly_fractions)
        assert stp.start_hour == st
        assert stp.end_hour == datetime.datetime(2015, 1, 1, 23)

    def test_two_days(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 3, 0)
        stp = StaticTimeProfiler(st, et, hourly_fractions=self.DAILY_HOURLY_FRACTIONS)

        # expected = 2* map(lambda x: x / 2,
        #     StaticTimeProfiler.DEFAULT_DAILY_HOURLY_FRACTIONS)
        # assert_approximately_equal(r, expected)
        assert stp.start_hour == st
        assert stp.end_hour == datetime.datetime(2015, 1, 2, 23)

        # stp = StaticTimeProfiler(self.DAILY_HOURLY_FRACTIONS)
        # r = stp._compute_hourly_fractions(s, e)
        # expected = 2* map(lambda x: x / 2, self.DAILY_HOURLY_FRACTIONS)
        # assert_approximately_equal(r, expected)
        # assert stp.start_hour == st
        # assert stp.end_hour == datetime.datetime(2015, 1, 2, 23)

        pass

    def test_partial_days(self):
        s = datetime.datetime(2015, 1, 1, 12, 20)
        e = datetime.datetime(2015, 1, 2, 16, 40)
        stp = StaticTimeProfiler(s, e,
            hourly_fractions=self.DAILY_HOURLY_FRACTIONS)
        # r = stp._compute_hourly_fractions(s, e)
        # print r
        # TODO: This is what is returned; verify manually that it is correct,
        # and if so, use it in an assert
        # [
        #     0.0,
        #     0.06411078343377354,
        #     0.08334401846390559,
        #     0.10257725349403765,
        #     0.108988331837415,
        #     0.07693294012052823,
        #     0.04487754840364148,
        #     0.025644313373509413,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.0036543146557250915,
        #     0.012822156686754706,
        #     0.025644313373509413,
        #     0.04487754840364148,
        #     0.06411078343377354,
        #     0.08334401846390559,
        #     0.10257725349403765,
        #     0.108988331837415
        # ]
        # TODO: add assert to check hourly fractions
        assert stp.start_hour == datetime.datetime(2015, 1, 1, 12)
        assert stp.end_hour == datetime.datetime(2015, 1, 2, 16)

class TestStaticTimeProfiler_CustomHourlyFractions(object):

    # 48 hours worth of hourly fractions
    HOURLY_FRACTIONS = {
        'area_fraction': [
            0.004, 0.006, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.10, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
            0.02, 0.02, 0.01, 0.00, 0.00, 0.00,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0
        ],
        'flaming': [
            0.005, 0.005, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.10, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
            0.02, 0.02, 0.01, 0.00, 0.00, 0.00,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0
        ],
        'residual': [  # shifted 6 hrs
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
            0.005, 0.005, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0
        ],
        'smoldering': [  # shifted another 6 hrss
            0.10, 0.10, 0.05, 0.05, 0.05, 0.05,
            0.05, 0.05, 0.05, 0.05, 0.03, 0.02,
            0.005, 0.005, 0.01, 0.01, 0.01, 0.01,
            0.05, 0.05, 0.05, 0.05, 0.05, 0.05,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0
        ]
    }

    def test_one_day(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 2, 0)
        with raises(InvalidHourlyFractionsError) as e:
            stp = StaticTimeProfiler(st, et, hourly_fractions=self.HOURLY_FRACTIONS)

    def test_two_days(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 3, 0)
        stp = StaticTimeProfiler(st, et, hourly_fractions=self.HOURLY_FRACTIONS)
        expected_hourly_fractions = {
            p: self.HOURLY_FRACTIONS[p] for p in stp.FIELDS
        }
        assert_approximately_equal(expected_hourly_fractions, stp.hourly_fractions)
        assert stp.start_hour == st
        assert stp.end_hour == datetime.datetime(2015, 1, 2, 23)

    def test_three_days(self):
        st = datetime.datetime(2015, 1, 1, 0)
        et = datetime.datetime(2015, 1, 4, 0)
        with raises(InvalidHourlyFractionsError) as e:
            stp = StaticTimeProfiler(st, et, hourly_fractions=self.HOURLY_FRACTIONS)


test/unit/timeprofile/test_base.py
__author__      = "Joel Dubowy"

import datetime

from py.test import raises

from timeprofile import (
    BaseTimeProfiler,
    InvalidStartEndTimesError
)

class TestBaseTimeProfiler_ValidationMethods(object):

    def test_validate_start_end_times(self, monkeypatch):
        profiler = BaseTimeProfiler()

        st = datetime.datetime(2015, 1, 1, 11)
        et = datetime.datetime(2015, 1, 1, 12)
        # start can't come after end
        with raises(InvalidStartEndTimesError) as e:
            profiler._validate_start_end_times(et, st)
        # start can't equal end
        with raises(InvalidStartEndTimesError) as e:
            profiler._validate_start_end_times(st, st)
        # proper order of dates shouldn't raise error
        profiler._validate_start_end_times(st, et)


timeprofile/__init__.py
__author__      = "Joel Dubowy"

__version_info__ = (1,1,2)
__version__ = '.'.join([str(n) for n in __version_info__])

import datetime

__all__ = [
    'BaseTimeProfiler',
    'InvalidStartEndTimesError'
]

class InvalidStartEndTimesError(ValueError):
    pass

class BaseTimeProfiler(object):

    ONE_HOUR = datetime.timedelta(hours=1)
    FIELDS = ['area_fraction', 'flaming', 'smoldering', 'residual']

    def _validate_start_end_times(self, local_start_time, local_end_time,
            time_qualifier=""):
        """Raises an InvalidStartEndTimesError exception if times are invalid.
        """
        # TODO: other checks ?
        if local_start_time >= local_end_time:
            raise InvalidStartEndTimesError("The fire's {} start time, {},"
                " is not before its end time, {}".format(time_qualifier,
                local_start_time.isoformat(), local_end_time.isoformat()))


timeprofile/static.py
"""timeprofile.static

In this simple, static time-profiling model, no environmental conditions
(such as reltative humidity) are considered.
"""

__author__      = "Joel Dubowy"

import datetime
from collections import defaultdict

from nested_dict import nested_dict
from functools import reduce

from . import BaseTimeProfiler, InvalidStartEndTimesError

__all__ = [
    'StaticTimeProfiler',
    'InvalidHourlyFractionsError'
]

class InvalidHourlyFractionsError(ValueError):
    pass

class StaticTimeProfiler(BaseTimeProfiler):

    DEFAULT_DAILY_HOURLY_FRACTIONS = defaultdict(lambda: [
        0.005700, # 00:00 (local time)
        0.005700, # 01:00
        0.005700, # 02:00
        0.005700, # 03:00
        0.005700, # 04:00
        0.005700, # 05:00
        0.005700, # 06:00
        0.005700, # 07:00
        0.005700, # 08:00
        0.005800, # 09:00
        0.020000, # 10:00
        0.040000, # 11:00
        0.070000, # 12:00
        0.100000, # 13:00
        0.130000, # 14:00
        0.160000, # 15:00
        0.170000, # 16:00
        0.120000, # 17:00
        0.070000, # 18:00
        0.040000, # 19:00
        0.005800, # 20:00
        0.005700, # 21:00
        0.005700, # 22:00
        0.005700 # 23:00
    ])
    """DEFAULT_DAILY_HOURLY_FRACTIONS will return the same hourly fractions
    for 'flaming', 'smoldering' and 'residual'

    These hourly fractions were copied from the wrap_time_profile module in
    BlueSky Framework, and modified slightly.  The 09:00 and 20:00 were both
    increased from 0.0057 to 0.0058 so that the values add up to 1.00.  (The
    values in wrap_time_profile add up to 0.9998)
    """

    def __init__(self, local_start_time, local_end_time, hourly_fractions=None): #, **options):
        """StaticTimeProfiler constructor

        kwargs:
         - hourly_fractions - custom hourly fractions of emissions; can be
           specified for all hours from start to end times, or for a single 24
           hour day, to be repeated or truncated to fill the time window defined
           by local_start_time / local_end_time

        *Note*: If len(hourly_fractions) == 24, they are assumed to represent
           fractions from 00:00 through 23:00.  If the start/end define a 24 hour
           period that does *not* start at 00:00, be aware that the values
           in hourly_fractions will *not* be applied from index 0 through 23.
           Whenever len(hourly_fractions) == 24 (regardless of the start/end),
           the local hour of day is used as the index to hourly_fractions.
        """
        # _compute_hourly_fractions will set self.start_hour, self.end_hour,
        # and self.hourly_fractions
        self._set_times(local_start_time, local_end_time)
        self._compute_hourly_fractions(hourly_fractions)

    ##
    ## Validation Methods
    ##

    def _validate_hourly_fractions(self, num_hours, hourly_fractions):
        """Raises an InvalidHourlyFractionsError exception if validation
        fails.
        """
        if hourly_fractions:
            for k in self.FIELDS:
                if (len(hourly_fractions.get(k, [])) not in (24, num_hours) or
                        abs(1 - sum(hourly_fractions[k])) > 0.001):
                    raise InvalidHourlyFractionsError(
                        "There must be 24 or {} hourly fractions that sum to 1.00"
                        " for each of the '{}' fields".format(
                        num_hours, ', '.join(self.FIELDS)))


    ##
    ## Computing Hourly Fractions
    ##


    def _set_times(self, local_start_time, local_end_time):
        """Sets start and end times, and start/end hour offsets

        Args:
         - local_start_time --
         - local_end_time --
        """
        self._validate_start_end_times(local_start_time, local_end_time)

        self._first_hour_offset = datetime.timedelta(
            minutes=local_start_time.minute, seconds=local_start_time.second)
        self.start_hour = local_start_time - self._first_hour_offset

        if local_end_time.time() == datetime.time(local_end_time.hour):
            # fire ended exactly on the hour
            self.end_hour = local_end_time - self.ONE_HOUR
            self._last_hour_offset = self.ONE_HOUR
        else:
            self._last_hour_offset = datetime.timedelta(
                minutes=local_end_time.minute, seconds=local_end_time.second)
            self.end_hour = local_end_time - self._last_hour_offset

    def _compute_hourly_fractions(self, hourly_fractions):
        """Determines what fraction of the fire's emissions occur in each
        calendar hour of the fire's duration.

        For example, if....

        Args:
         - hourly_fractions --
        """

        # TODO: use math.ceil instead of int? (should have divided evenly, so prob not)
        num_hours = int((self.end_hour - self.start_hour).total_seconds() / 3600) + 1
        self._validate_hourly_fractions(num_hours, hourly_fractions)

        hourly_fractions = hourly_fractions or self.DEFAULT_DAILY_HOURLY_FRACTIONS

        # TODO: if more efficient, iterate by phase within the while loop
        new_hourly_fractions = {}
        for p in self.FIELDS:
            num_hourly_fractions = len(hourly_fractions[p])
            r = []
            i = 0
            d = self.start_hour
            while d <= self.end_hour:
                idx = d.hour if num_hourly_fractions == 24 else i
                f = hourly_fractions[p][idx]
                if d == self.start_hour:
                    f *= (3600 - self._first_hour_offset.seconds) / 3600
                elif d == self.end_hour:
                    f *= self._last_hour_offset.seconds / 3600
                r.append(f)
                d += self.ONE_HOUR
                i += 1

            # Normalize so that it all adds up to 1.0
            total = reduce(lambda x, y: x + y, r)
            new_hourly_fractions[p] = [x / total for x in r]

        self.hourly_fractions = new_hourly_fractions
